Adds support for waves functions

Rogerrrrrrrs opened the pull request:
This pull request adds Wave functionality and resolves #38 and closes #64 and is basically a continuation of #64  

It also adds ticks to the interrupt event as discussed [here](https://github.com/fivdi/pigpio/pull/64#issuecomment-433185029)

The following methods where added:
* waveClear
* waveAddNew
* waveAddGeneric
* waveCreate
* waveDelete
* waveTxSend
* waveChain
* waveTxAt
* waveTxBusy
* waveTxStop
* waveGetMicros
* waveGetHighMicros
* waveGetMaxMicros
* waveGetPulses
* waveGetHighPulses
* waveGetMaxPulses
* waveGetCbs
* waveGetHighCbs
* waveGetMaxCbs

The wave methods where tested using a Raspberry Pi 3B+ and verified using an oscilloscope.

fivdi commented:
@Rogerrrrrrrs thank you for submitting this PR and all the work you have done on it, @AnthonyMyatt, @bahamut657 and thank you for your work on this PR too.

@Rogerrrrrrrs 

All the new wave methods are currently methods in [class Gpio](https://github.com/Rogerrrrrrrs/pigpio/blob/Waves/doc/gpio.md#class-gpio---general-purpose-input-output). Why is this the case? Would it not be better if they were functions in [global](https://github.com/Rogerrrrrrrs/pigpio/blob/Waves/doc/global.md)?

When the PR is installed with Node.js v13.5.0 using the command `npm i https://github.com/Rogerrrrrrrs/pigpio.git#Waves` there are compile errors that need to be fixed.

<details><summary>Node.js v13.5.0 Compile Errors</summary>
<p>

```
pi@raspberrypi:~/pigpio $ npm i https://github.com/Rogerrrrrrrs/pigpio.git#Waves
 

> pigpio@2.0.1 install /home/pi/pigpio/node_modules/pigpio
> node-gyp rebuild

make: Entering directory '/home/pi/pigpio/node_modules/pigpio/build'
  CXX(target) Release/obj.target/pigpio/src/pigpio.o
../src/pigpio.cc: In function ‘Nan::NAN_METHOD_RETURN_TYPE gpioWaveAddGeneric(Nan::NAN_METHOD_ARGS_TYPE)’:
../src/pigpio.cc:746:50: error: no matching function for call to ‘v8::Array::Get(unsigned int&)’
       v8::Local<v8::Value> element = array->Get(i);
                                                  ^
In file included from /home/pi/.cache/node-gyp/13.5.0/include/node/node.h:63,
                 from ../../nan/nan.h:54,
                 from ../src/pigpio.cc:3:
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
../src/pigpio.cc:756:90: error: no matching function for call to ‘v8::Object::Get(v8::MaybeLocal<v8::String>)’
 cal < v8::Value > _on = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOn"));
                                                                              ^

In file included from /home/pi/.cache/node-gyp/13.5.0/include/node/node.h:63,
                 from ../../nan/nan.h:54,
                 from ../src/pigpio.cc:3:
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
../src/pigpio.cc:757:92: error: no matching function for call to ‘v8::Object::Get(v8::MaybeLocal<v8::String>)’
 l < v8::Value > _off = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOff"));
                                                                              ^

In file included from /home/pi/.cache/node-gyp/13.5.0/include/node/node.h:63,
                 from ../../nan/nan.h:54,
                 from ../src/pigpio.cc:3:
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
../src/pigpio.cc:758:94: error: no matching function for call to ‘v8::Object::Get(v8::MaybeLocal<v8::String>)’
 < v8::Value > _delay = pulse->Get(v8::String::NewFromUtf8(isolate, "usDelay"));
                                                                              ^

In file included from /home/pi/.cache/node-gyp/13.5.0/include/node/node.h:63,
                 from ../../nan/nan.h:54,
                 from ../src/pigpio.cc:3:
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, v8::Local<v8::Value>)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3594:43: note:   candidate expects 2 arguments, 1 provided
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note: candidate: ‘v8::MaybeLocal<v8::Value> v8::Object::Get(v8::Local<v8::Context>, uint32_t)’
   V8_WARN_UNUSED_RESULT MaybeLocal<Value> Get(Local<Context> context,
                                           ^~~
/home/pi/.cache/node-gyp/13.5.0/include/node/v8.h:3597:43: note:   candidate expects 2 arguments, 1 provided
make: *** [pigpio.target.mk:111: Release/obj.target/pigpio/src/pigpio.o] Error 1
make: Leaving directory '/home/pi/pigpio/node_modules/pigpio/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/pi/opt/node-v13.5.0-linux-armv7l/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:194:23)
gyp ERR! stack     at ChildProcess.emit (events.js:305:20)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:275:12)
gyp ERR! System Linux 4.19.75-v7l+
gyp ERR! command "/home/pi/opt/node-v13.5.0-linux-armv7l/bin/node" "/home/pi/opt/node-v13.5.0-linux-armv7l/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/pi/pigpio/node_modules/pigpio
gyp ERR! node -v v13.5.0
gyp ERR! node-gyp -v v5.0.5
gyp ERR! not ok 
npm WARN enoent ENOENT: no such file or directory, open '/home/pi/pigpio/package.json'
npm WARN pigpio No description
npm WARN pigpio No repository field.
npm WARN pigpio No README data
npm WARN pigpio No license field.

npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! pigpio@2.0.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the pigpio@2.0.1 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/pi/.npm/_logs/2020-01-10T23_08_34_457Z-debug.log
pi@raspberrypi:~/pigpio $ 
```

</p>
</details>

I'm in the process of reviewing the PR.

fivdi commented on pigpio.js:
```diff
@@ -103,7 +103,7 @@ class Gpio extends EventEmitter {
     return this;
   }
-
+  
```
It's not necessary to add a space to this line. It should be removed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,209 @@ NAN_METHOD(gpioNotifyClose) {
+    return ThrowPigpioError(rc, "gpioWaveClear");
+  }
+  
+  info.GetReturnValue().Set(rc);
```
`info.GetReturnValue().Set(rc);` is not needed and should be removed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,209 @@ NAN_METHOD(gpioNotifyClose) {
+    return ThrowPigpioError(rc, "gpioWaveAddNew");
+  }
+
+  info.GetReturnValue().Set(rc);
```
`info.GetReturnValue().Set(rc);` is not needed ans should be removed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,209 @@ NAN_METHOD(gpioNotifyClose) {
+    return ThrowPigpioError(rc, "gpioWaveDelete");
+  }
+
+  info.GetReturnValue().Set(rc);
```
`info.GetReturnValue().Set(rc);` is not needed and should be removed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,209 @@ NAN_METHOD(gpioNotifyClose) {
+    return ThrowPigpioError(rc, "gpioWaveChain");
+  }
+
+  info.GetReturnValue().Set(rc);
```
`info.GetReturnValue().Set(rc);` is not needed and should be removed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,209 @@ NAN_METHOD(gpioNotifyClose) {
+    return ThrowPigpioError(rc, "gpioWaveTxStop");
+  }
+
+  info.GetReturnValue().Set(rc);
```
`info.GetReturnValue().Set(rc);` is not needed and should be removed.

fivdi commented:
@Rogerrrrrrrs If you're not sure how to fix the compile errors that occur when installing with with Node.js v13.5.0 please let me know and I'll take a look. There are no compile errors with Node.js v8, v10 and v12 and everything functions correctly.

Rogerrrrrrrs commented:
@fivdi  Thank you, I would be happy to take your help on the compile errors.

`All the new wave methods are currently methods in class Gpio. Why is this the case? Would it not be better if they were functions in global?`

You are right, there is no benefit being a member of Gpio.

I agree with the requested changes and will take a look at it now.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,199 @@ NAN_METHOD(gpioNotifyClose) {
+  for (unsigned i = 0; i < array->Length(); i++) {
+    if (Nan::Has(array, i).FromJust()) {
+      
+      v8::Local<v8::Value> element = array->Get(i);
```
To fix the compile errors with Node.js v13.5.0 change this:
```cpp
       v8::Local<v8::Value> element = array->Get(i);
```
to this:
```cpp
      v8::Local<v8::Value> element = Nan::Get(array, i).ToLocalChecked();
```

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,199 @@ NAN_METHOD(gpioNotifyClose) {
+
+      v8::Local<v8::Object> pulse = v8::Local<v8::Object>::Cast(element);
+
+      v8::Isolate * isolate = v8::Isolate::GetCurrent();
```
`       v8::Isolate * isolate = v8::Isolate::GetCurrent();` can be removed as the code that is currently using `isolate` will be modified and `isolate` will no longer be needed.

fivdi commented on src/pigpio.cc:
```diff
@@ -707,6 +707,199 @@ NAN_METHOD(gpioNotifyClose) {
+
+      v8::Local < v8::Value > _on = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOn"));
+      v8::Local < v8::Value > _off = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOff"));
+      v8::Local < v8::Value > _delay = pulse->Get(v8::String::NewFromUtf8(isolate, "usDelay"));
```
To fix the compile errors with Node.js v13.5.0 change this:
```cpp
      v8::Local < v8::Value > _on = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOn"));
      v8::Local < v8::Value > _off = pulse->Get(v8::String::NewFromUtf8(isolate, "gpioOff"));
      v8::Local < v8::Value > _delay = pulse->Get(v8::String::NewFromUtf8(isolate, "usDelay"));
```
to this:
```cpp
      v8::Local<v8::Value> _on =
        Nan::Get(pulse, Nan::New<v8::String>("gpioOn").ToLocalChecked()).
        ToLocalChecked();
      v8::Local<v8::Value> _off =
        Nan::Get(pulse, Nan::New<v8::String>("gpioOff").ToLocalChecked()).
        ToLocalChecked();
      v8::Local<v8::Value> _delay =
        Nan::Get(pulse, Nan::New<v8::String>("usDelay").ToLocalChecked()).
        ToLocalChecked();
```

fivdi commented:
@Rogerrrrrrrs please let me know when you think you're finished.

Rogerrrrrrrs commented:
@fivdi I have moved the wave functions to the global level. I have updated the documentation, examples and tests. When I did that, I also had to update the return values of the functions that previously returned `this`.
It also builds now without errors on node 13 thanks to your help.
I think you can review it again.

fivdi commented:
@Rogerrrrrrrs , @bahamut657, @AnthonyMyatt Thank you all very much for working on the PR.

AnthonyMyatt commented:
Glad I could play a part.

fivdi commented:
This PR has now been published with pigpio@3.0.0 on npm.

cinderblock commented:
@fivdi Were there any breaking changes in the 3.0.0 release? What about the 2.0.0 release?

fivdi commented:
There were no breaking changes in v2.0.0 or v3.0.0.
